diff -ruN mz-0.40/src/dns.c mz-patched/src/dns.c
--- mz-0.40/src/dns.c	2009-08-19 22:47:01.000000000 +0300
+++ mz-patched/src/dns.c	2014-12-13 12:24:11.888869254 +0200
@@ -26,7 +26,7 @@
 
 #include "mz.h"
 #include "cli.h"
-
+#include <sys/time.h>
 
 #define MZ_DNS_HELP \
    		"| DNS type: Send Domain Name System Messages.\n" \
@@ -90,13 +90,13 @@
 		"\n"
 
 
-int dns_get_query (char* argval);
+int dns_get_query (char* argval, int type);
 int dns_get_answer (char* argval);
 
 
 
 // Note: I do NOT use libnet here (had problems with bugs there...)
-int create_dns_packet ()
+int create_dns_packet (int dns_type)
 {
    
    char  *token, *tokenptr, argval[MAX_PAYLOAD_SIZE];
@@ -184,15 +184,15 @@
    /////////////////////////////////////////////////////////////////////////////////
    // Evaluate CLI parameters:
 
-   
+      
    //  Handle the query //
 
-   if ( (getarg(tx.arg_string,"query", argval)==1) ||
-	(getarg(tx.arg_string,"q", argval)==1) )
-     {
+   if ( dns_type == 1)
+   {
+      if ( (getarg(tx.arg_string,"query", argval)==1) || (getarg(tx.arg_string,"q", argval)==1) )
+      {
+	(void) dns_get_query (argval, dns_type); // returns the length in byte	dns_num_q0=1;
 
-	(void) dns_get_query (argval); // returns the length in byte	dns_num_q0=1;
-	
 	// copy the result from gbuf to our local buffer 'section':
 	for (j=0;j<gbuf_s;j++)
 	  {
@@ -221,7 +221,40 @@
 	dns_flags_rcode        = 0; // no errors
 	dns_num_q0             = 1; // number of questions
      }
+   }
+   else
+   {
+	(void) dns_get_query (argval, dns_type); // returns the length in byte dns_num_q0=1;
+
+        // copy the result from gbuf to our local buffer 'section':
+        for (j=0;j<gbuf_s;j++)
+          {
+             section[j]=gbuf[j];
+          }
+
+        i = gbuf_s;
+
+        // Set defaults if not already set by callee.
+        // !! But ONLY set these if there is no additional answer section
+        // !! because then the answer section should set the defaults !!!
+        if ( (getarg(tx.arg_string,"answer", NULL)==0) && // no answer
+             (getarg(tx.arg_string,"a", NULL)==0) )
+          {
+             if (!tx.dp)     tx.dp = 53;
+             if (!tx.sp)     tx.sp = 42000;
+          }
+
+
+        // These are the defaults for a query:
+        dns_flags_aa           = 1; // authoritative answer
+        dns_flags_tc           = 0; // not truncated
+        dns_flags_rd           = 1; // recursion desired
+        dns_flags_ra           = 0; // recursion available
+        dns_flags_z            = 0; // FYI: if 010 = 2 = authenticated
+        dns_flags_rcode        = 0; // no errors
+        dns_num_q0             = 1; // number of questions
 
+   }
 
    
    // Handle the answer:
@@ -525,6 +558,27 @@
 }
 
 
+// galkan
+long ms(void) 
+{
+    struct timeval te; 
+	
+    gettimeofday(&te, NULL); 
+    long milliseconds = te.tv_sec*1000L + te.tv_usec/1000; // caculate milliseconds
+    return milliseconds;
+}
+
+void gen_random(char *s, const int len) 
+{
+	int i;
+	static const char alphanum[] = "abcdefghijklmnopqrstuvwxyz";
+
+	for (i = 0; i < len - 1; ++i) 
+		s[i] = alphanum[rand() % (sizeof(alphanum)-1)];
+
+	s[i] = '\0';
+}
+// galkan
 
 ////////////////////////////////////////////////////////////////////////////////////////////
 // Accepts a string like "www.perihel.at:A" or "www.perihel.at"
@@ -534,8 +588,9 @@
 // Return value: 
 //   number of queries (currently only 1 query accepted, 
 //   hence return value is 1 on success or 0 upon failure
-//   
-int dns_get_query(char* argval)
+//  
+ 
+int dns_get_query(char* argval, int type)
 {
    char *token, *field, *saveptr1=NULL, *saveptr2=NULL;
    int i,j, cnt;
@@ -543,12 +598,59 @@
    unsigned char *x;
    
    i=0;
-   
+  
+   srand(ms()); 
+
+   if ( type == 1 ) {	
+   	char *new_argval = (char *)(malloc(512));
+	if (new_argval == NULL) {
+		fprintf(stderr, "Cannot allocate memory !!!\n");
+		exit(EXIT_FAILURE);
+	}
+
+   	gen_random(new_argval, 6);
+   	strncat(new_argval, ".", 1);	
+   	strncat(new_argval, argval, strlen(argval));
+   	strncpy(argval, new_argval, strlen(new_argval));
+
+	free(new_argval);	
+   } else {
+	char *new_argval = (char *)(malloc(512));
+	if (new_argval == NULL) {
+                fprintf(stderr, "Cannot allocate memory !!!\n");
+                exit(EXIT_FAILURE);
+        }
+
+	char *tmp = (char *)(malloc(512));
+	if (tmp == NULL) {
+                fprintf(stderr, "Cannot allocate memory !!!\n");
+                exit(EXIT_FAILURE);
+        }
+	memset(new_argval, 0x0, 512);
+
+        gen_random(tmp, 4);
+	strncat(tmp, ".", 1);
+	strncat(new_argval, tmp, strlen(tmp));
+	
+	gen_random(tmp, 9);
+	strncat(tmp, ".", 1);
+	strncat(new_argval, tmp, strlen(tmp));
+
+	gen_random(tmp, 4);
+	strncat(new_argval, tmp, strlen(tmp));
+
+	strncpy(argval, new_argval, strlen(new_argval));	
+
+        free(new_argval);
+	free(tmp);
+   }
+
    // now get first field: <name>
    field = strtok_r(argval, ":", &saveptr1);
    
    // decompose <name> into labels:
    token = strtok_r(field, ".", &saveptr2);
+ 
 
    do   // loop through all labels 
      {
diff -ruN mz-0.40/src/init.c mz-patched/src/init.c
--- mz-0.40/src/init.c	2010-07-07 17:31:10.000000000 +0300
+++ mz-patched/src/init.c	2014-12-13 11:46:28.020997119 +0200
@@ -212,7 +212,7 @@
 	opterr = 1; // let getopt print error message if necessary
 	
 	
-	while ((c = getopt (argc, argv, "hqvVSxra:A:b:B:c:d:f:F:p:P:t:T:M:Q:X:")) != -1)
+	while ((c = getopt (argc, argv, "hqvVSxra:A:b:B:c:d:f:F:p:P:t:T:M:Q:X:g:")) != -1)
 		switch (c) {
 		 case 'h':
 			usage();
@@ -290,6 +290,10 @@
 		 case 't':
 			packet_type = optarg; // analyzed below
 			break;
+		 case 'g':
+			strcpy(optarg, "random_dns");
+                        packet_type = optarg; // analyzed below
+                        break;
 		 case 'X':
 			mops_type = optarg; // MOPS TRANSITION STRATEGY -- analyzed below
 			break;
@@ -475,17 +479,19 @@
 		
 		// Set destination IP address:
 		if (strlen(tx.ip_dst_txt)) {  // option -B has been specified
-			if (mz_strcmp(tx.ip_dst_txt, "rand", 2)==0) {
-				fprintf(stderr, "Option -B does not support random destination IP addresses currently.\n");
-				return 1;
-			}
+			if(access( tx.ip_dst_txt, F_OK ) == -1) { 
+				if (mz_strcmp(tx.ip_dst_txt, "rand", 2)==0) {
+					fprintf(stderr, "Option -B does not support random destination IP addresses currently.\n");
+					return 1;
+				}
 		  
-			if (mz_strcmp(tx.ip_dst_txt, "bcast", 2)==0) {
-				tx.ip_dst = libnet_name2addr4 (l, "255.255.255.255", LIBNET_DONT_RESOLVE);	
-			} else if (get_ip_range_dst(tx.ip_dst_txt)) { // returns 1 when no range has been specified
-				// name2addr4 accepts a DOTTED DECIMAL ADDRESS or a FQDN:
-				tx.ip_dst = libnet_name2addr4 (l, tx.ip_dst_txt, LIBNET_RESOLVE);		     
-			}
+				if (mz_strcmp(tx.ip_dst_txt, "bcast", 2)==0) {
+					tx.ip_dst = libnet_name2addr4 (l, "255.255.255.255", LIBNET_DONT_RESOLVE);	
+				} else if (get_ip_range_dst(tx.ip_dst_txt)) { // returns 1 when no range has been specified
+					// name2addr4 accepts a DOTTED DECIMAL ADDRESS or a FQDN:
+					tx.ip_dst = libnet_name2addr4 (l, tx.ip_dst_txt, LIBNET_RESOLVE);		     
+				}
+			} 
 		}
 		else { // no destination IP specified: by default use broadcast
 			tx.ip_dst = libnet_name2addr4 (l, "255.255.255.255", LIBNET_DONT_RESOLVE);	
@@ -575,7 +581,7 @@
 		}
 	}
 	
-	
+
    if (packet_type == NULL) { // raw hex string given
 	   mode = BYTE_STREAM;
    }
@@ -600,6 +606,9 @@
 	else if (strcmp(packet_type,"dns")==0) {
 		mode = DNS;
 	}
+	else if (strcmp(packet_type,"random_dns")==0) {
+                mode = RANDOM_DNS;
+        }
 	else if (strcmp(packet_type,"cdp")==0) {
 		mode = CDP;
 	}
Binary files mz-0.40/src/mz and mz-patched/src/mz differ
diff -ruN mz-0.40/src/mz.c mz-patched/src/mz.c
--- mz-0.40/src/mz.c	2010-06-02 00:06:21.000000000 +0300
+++ mz-patched/src/mz.c	2014-12-13 11:46:28.020997119 +0200
@@ -127,7 +127,7 @@
 		   "|                        encapsulations in lower layers. Most packet types allow/require\n"
 		   "|                        additional packet-specific arguments in an arg_string.\n"
 		   "|                        Currently supported types: arp, bpdu, cdp, ip, icmp, udp, tcp,\n"
-		   "|                        dns, rtp, syslog, lldp.\n"
+		   "|                        dns, rtp, syslog, lldp, random_dns.\n"
 		   "|                        For context-help use 'help' as arg_string!\n"
 		   "|  -T <packet_type>      Specify packet type for server mode. Currently only rtp is supported.\n"
 		   "|                        Enter -T help or -T rtp help for further information.\n"
@@ -148,13 +148,19 @@
 int main(int argc, char *argv[])
 {
 
-   
+   // galkan
+	FILE *fp;
+	char dst_line[1024];
+   // galkan
+
    // These handles are only used when creating L3 and above packets.
    libnet_t             *l;               // the context 
    libnet_ptag_t         t2=0, t3=0, t4=0;      // handles to layers 
    
    double cpu_time_used;
 
+   int sayac, dns_type;
+
    // Check if we have root priviliges
    if ( (getuid()!=0) && (geteuid()!=0) )
      {
@@ -250,6 +256,7 @@
 	  send_frame (l, t3, t4); // NOTE: send_frame also destroys context finaly
 	break;
 	
+
       case TCP:
 	tx.ip_proto = 6;    
 	l = get_link_context();
@@ -261,20 +268,89 @@
 	else
 	  send_frame (l, t3, t4); // NOTE: send_frame also destroys context finaly
 	break;
-	
+
+
+
       case DNS:
 	tx.ip_proto = 17;
-	l = get_link_context();
-	(void) create_dns_packet();
-	t4 = create_udp_packet(l);     // t4 can be used for later header changes
-	t3 = create_ip_packet(l);      // t3 can be used for later header changes
-	if (!quiet) complexity();
-	if (tx.packet_mode==0)         // Ethernet manipulation features does NOT use ARP to determine eth_dst  
-	  t2 = create_eth_frame(l, t3, t4);    // t2 can be used for later header changes
-	else
-	  send_frame (l, t3, t4); // NOTE: send_frame also destroys context finaly
-	break;
-	
+	dns_type = 1;
+	for (sayac=0; sayac<tx.count; sayac++)
+	{
+		l = get_link_context();
+		(void) create_dns_packet(dns_type);
+		t4 = create_udp_packet(l);     // t4 can be used for later header changes
+		t3 = create_ip_packet(l);      // t3 can be used for later header changes
+		if (!quiet) complexity();
+
+		if (tx.packet_mode==0)         // Ethernet manipulation features does NOT use ARP to determine eth_dst  
+	  		t2 = create_eth_frame(l, t3, t4);    // t2 can be used for later header changes
+		else {
+			if(access( tx.ip_dst_txt, F_OK ) != -1) {
+				fp = fopen(tx.ip_dst_txt, "r");
+                               	if (fp == NULL) {
+                                        fprintf(stderr, "File: %s cannot be opened !!!", tx.ip_dst_txt);
+                                        exit(EXIT_FAILURE);
+                               	}
+			
+				while ( fgets ( dst_line, sizeof dst_line, fp ) != NULL ) {
+					dst_line[strlen(dst_line)-1] = '\0';
+					tx.ip_dst = libnet_name2addr4 (l, dst_line, LIBNET_DONT_RESOLVE);
+					l = get_link_context();
+					(void) create_dns_packet(dns_type);
+					t4 = create_udp_packet(l);
+					t3 = create_ip_packet(l);
+					
+					send_frame (l, t3, t4);
+				}
+				fclose(fp);	
+			} else
+				send_frame (l, t3, t4);
+		}
+	}
+	break;
+
+
+
+      case RANDOM_DNS:
+        tx.ip_proto = 17;
+	dns_type = 2;
+        for (sayac=0; sayac<tx.count; sayac++)
+        {
+                l = get_link_context();
+                (void) create_dns_packet(dns_type);
+                t4 = create_udp_packet(l);     // t4 can be used for later header changes
+                t3 = create_ip_packet(l);      // t3 can be used for later header changes
+                if (!quiet) complexity();
+
+                if (tx.packet_mode==0)         // Ethernet manipulation features does NOT use ARP to determine eth_dst
+                        t2 = create_eth_frame(l, t3, t4);    // t2 can be used for later header changes
+                else {
+                        if(access( tx.ip_dst_txt, F_OK ) != -1) {
+                                fp = fopen(tx.ip_dst_txt, "r");
+                                if (fp == NULL) {
+                                        fprintf(stderr, "File: %s cannot be opened !!!", tx.ip_dst_txt);
+                                        exit(EXIT_FAILURE);
+                                }
+
+                                while ( fgets ( dst_line, sizeof dst_line, fp ) != NULL ) {
+                                        dst_line[strlen(dst_line)-1] = '\0';
+                                        tx.ip_dst = libnet_name2addr4 (l, dst_line, LIBNET_DONT_RESOLVE);
+                                        l = get_link_context();
+                                        (void) create_dns_packet(dns_type);
+                                        t4 = create_udp_packet(l);
+                                        t3 = create_ip_packet(l);
+
+                                        send_frame (l, t3, t4);
+                                }
+                                fclose(fp);
+                        } else
+                                send_frame (l, t3, t4);
+		}
+        }
+        break;
+
+
+
       case RTP:
 	tx.ip_proto = 17;
 	l = get_link_context();
diff -ruN mz-0.40/src/mz.h mz-patched/src/mz.h
--- mz-0.40/src/mz.h	2010-06-22 01:10:59.000000000 +0300
+++ mz-patched/src/mz.h	2014-12-13 11:46:28.024997119 +0200
@@ -247,7 +247,8 @@
      RTP,
      RX_RTP,
      SYSLOG,
-     LLDP
+     LLDP,
+     RANDOM_DNS	
 } mode;
 
 
@@ -521,6 +522,7 @@
 libnet_t*      get_link_context();
 libnet_ptag_t  create_ip_packet (libnet_t *l);
 int            send_frame (libnet_t *l, libnet_ptag_t  t3, libnet_ptag_t  t4);
+int            send_dns_frame (libnet_t *l, libnet_ptag_t  t3, libnet_ptag_t  t4);
 
 
 
@@ -539,7 +541,7 @@
 //   Prototypes: Layer 7
 //
 // ************************************
-int  create_dns_packet ();
+int  create_dns_packet (int control);
 int  create_rtp_packet();
 int create_syslog_packet();
 
diff -ruN mz-0.40/src/send.c mz-patched/src/send.c
--- mz-0.40/src/send.c	2009-08-25 22:02:02.000000000 +0300
+++ mz-patched/src/send.c	2014-12-13 11:46:28.024997119 +0200
@@ -131,12 +131,13 @@
      {
 	cli_print(gcli, "Mausezahn will send %.Lf frames...\r", total_d);
      }
-   else
-     {
-	fprintf(stderr, "Mausezahn will send %.Lf frames... ", total_d);
-	fflush(stderr);
-	if (verbose) fprintf(stderr,"\n");
-     }
+   //else
+     //{
+	//fprintf(stderr, "Mausezahn will send %.Lf frames... ", total_d);
+	//fprintf(stderr, "... ");
+	//fflush(stderr);
+	//if (verbose) fprintf(stderr,"\n");
+    // }
    
    
    
@@ -262,3 +263,102 @@
    return 0;
 }
 
+
+
+///////////////////////////////////////////////////////////////////////
+//
+// Send complete frame (layers 2, 3, 4) multiple times if required
+// 
+// 
+int send_dns_frame (libnet_t *l, libnet_ptag_t  t3, libnet_ptag_t  t4)
+{
+   int count;
+
+   int // local vars are faster ;-)
+     tcp_seq_delta,
+     dp_isrange,
+     sp_isrange,
+     ip_dst_isrange,
+     ip_src_isrange,
+     rtp_mode=0;
+
+   
+   count = tx.count;
+   tcp_seq_delta = tx.tcp_seq_delta;
+   dp_isrange = tx.dp_isrange;
+   sp_isrange = tx.sp_isrange;
+   ip_dst_isrange = tx.ip_dst_isrange;
+   ip_src_isrange = tx.ip_src_isrange | tx.ip_src_rand;
+   if (mode == RTP) rtp_mode = 1;
+   
+   //if (count==0) goto AGAIN;
+   //for (i=0; i<count; i++)
+   //{
+
+	AGAIN:
+
+	if (verbose) (void) print_frame_details();
+	libnet_write(l);
+	if (mz_rand) tx.delay=(unsigned int) tx.delay*rand()/RAND_MAX;
+	if (tx.delay) SLEEP (tx.delay);
+	
+	//   No layer-2 modifications done here 
+	//   (see create_eth_frame which does L2 modifications additionally)
+
+	
+	if (tcp_seq_delta)
+	  {
+	     if (update_TCP_SQNR(l, t4)==0) // end of range not yet reached
+	       {
+		  goto AGAIN;
+	       }
+	  }
+	
+	if (dp_isrange)
+	  {
+	     if (update_DPORT(l, t4)==0) // end of range not yet reached
+	       {
+		  goto AGAIN;
+	       }
+	  }
+	
+	if (sp_isrange)
+	  {
+	     if (update_SPORT(l, t4)==0) // end of range not yet reached
+	       {
+		  goto AGAIN;
+	       }
+	  }
+	
+
+	if (ip_dst_isrange)
+	  {
+	     if (update_IP_DA(l, t3)==0) // end of range not yet reached
+	       {
+		  goto AGAIN;
+	       }
+	  }
+	
+	if (ip_src_isrange) // also catches random SA (see above)
+	  {
+	     if (update_IP_SA(l, t3)==0) // end of range not yet reached
+	       {
+		  goto AGAIN;
+	       }
+	  }
+
+	if (rtp_mode) // update SQNR and Timestamps in RTP header and payload
+	  {
+	     update_RTP(l, t4);
+	  }
+	
+	
+	if (!count) goto AGAIN;
+     //}
+   
+   libnet_destroy(l);
+   
+   return 0;
+}
+
+
diff -ruN mz-0.40/src/tx_switch.c mz-patched/src/tx_switch.c
--- mz-0.40/src/tx_switch.c	2010-03-07 14:44:55.000000000 +0200
+++ mz-patched/src/tx_switch.c	2014-12-13 11:46:28.024997119 +0200
@@ -100,7 +100,7 @@
       case DNS:
 	tx.ip_proto = 17;
 	l = get_link_context();
-	if (create_dns_packet()==-1) return 0;
+	if (create_dns_packet(1)==-1) return 0;
 	t4 = create_udp_packet(l);     // t4 can be used for later header changes
 	t3 = create_ip_packet(l);      // t3 can be used for later header changes
 	if (!quiet) complexity();
